File: UndirectedCycle.java
import java.util.HashSet;
import java.util.Set;

public class UndirectedCycle {
    private Graph g;
    private Set<String> marked;
    private Set<String> edges;
    private boolean hasCycle;

    public UndirectedCycle(Graph g) {
        this.g = g;
        marked = new HashSet<>();
        edges = new HashSet<>();
        hasCycle = false;
        for (String v : g.getVerts()) {
            if (!marked.contains(v)) {
                System.out.println("Testando " + v);
                hasCycle = containsCycle(v);
                if (hasCycle)
                    break;
            }
        }
    }

    public boolean containsCycle() {
        return hasCycle;
    }

    private boolean containsCycle(String v) {
        System.out.println("estou em " + v);
        marked.add(v);
        for (String w : g.getAdj(v)) {
            String e = v + "-" + w;
            System.out.println("e: " + e);
            if (!marked.contains(w)) {
                edges.add(e);
                boolean res = containsCycle(w);
                if (res)
                    return res;
            } else if (!edges.contains(w + "-" + v))
                return true;
        }
        return false;
    }
}

File: undirectedcycle.cpp
#include "undirectedcycle.h"
#include <iostream>
#include <fstream>
#include <sstream>
#include <set>

using namespace std;

UndirectedCycle::UndirectedCycle(Graph &g)
{
	hasCycle = false;
	for (auto const &w : g.getVerts())
	{
		if (marked.find(w) == marked.end())
		{
			hasCycle = containsCycle(g, w);
			if (hasCycle)
				break;
		}
	}
}

bool UndirectedCycle::containsCycle()
{
	return hasCycle;
}

bool UndirectedCycle::containsCycle(Graph &g, string v)
{
	cout << "Estou em " << v << endl;
	marked.insert(v);
	for (auto const &w : g.getAdj(v))
	{
		string e = v + "-" + w;
		if (marked.find(w) == marked.end())
		{
			edges.insert(e);
			bool res = containsCycle(g, w);
			if (res)
				return res;
		}
		else if (edges.find(w + "-" + v) == edges.end())
			return true;
	}
	return false;
}
